src/software/geom/util.cpp

```
@@ -1251,6 +1116,172 @@ std::optional<Segment> mergeFullyOverlappingSegments(Segment segment1, Segment s
     }
 }
 
+std::optional<std::vector<Segment>> reduceParallelSegments(std::vector<Segment> segments)
```

Is there any reason this needs to return an `optional`? Seems  like an empty vector would be fine if an empty vector is given or if no segments are collinear.

---

That's a very good point.

Changed.

---

Why would an empty vector be fine? Consider the case where this is given 5 vectors, the first 2 are parallel with each other, the last 3 are parallel with each other, but the first 2 are not parallel with the last 3. What's the expected behavior? I would much rather we throw an exception if we violate a precondition of this function (or I think Mathew suggested somewhere else that we project onto a segment and then do the reduction?)